### PR TITLE
Document variable substitutions in hidden fields in forms

### LINKED ISSUE #6420: Document var-subs in amp-form
Follow up to #5654 and #6130

### PR DESCRIPTION
**Pull request by mkhatib**
Closes #6420

### CODE REVIEW

**Comment by rudygalfi** [comment]
A couple comments, one about implementation (sorry).

cc @bpaduch to make sure we're accounting for form analytics as part of our overall analytics documentation (along with other specialized cases like amp-carousel, amp-access, etc).

**Comment by mkhatib** [comment]
@rudygalfi You might need to submit your review for me to see the comments.

File: extensions/amp-form/amp-form.md [hunk 1]
```diff
@@ -291,6 +296,38 @@ This shows all validation errors on all invalid inputs when the form is submitte
 #### As You Go
 This allows your user to see validation messages as they're interacting with the input, if the email they typed is invalid they'll see the error right away and once fixed the error goes away.
 
+## Variable Substitutions
+__(<a href="https://www.ampproject.org/docs/reference/experimental.html">experimental</a>)__
+
+`amp-form` allows [variable substitutions](../../spec/amp-var-substitutions.md) for inside hidden inputs with `default-value` attribute. On each submission, `amp-form` finds all `input[type=hidden]` inside the form and apply variable substitutions to `default-value` attribute and fill the result in the input `value`.
```
Comment (bpaduch) [inline extensions/amp-form/amp-form.md]: Rewrite as follows:

`amp-form` allows [variable substitutions](../../spec/amp-var-substitutions.md) for inputs that are `hidden` and that have the  `default-value` attribute.  On each form submission, `amp-form` finds all `input[type=hidden]` inside the form and applies variable substitutions to the `default-value` attribute and fills in the result for the input `value`.

File: extensions/amp-form/amp-form.md [hunk 2]
```diff
@@ -23,7 +23,12 @@ limitations under the License.
   </tr>
   <tr>
     <td width="40%"><strong>Availability</strong></td>
-    <td>Stable<br>(<a href="#custom-validations">Custom Validation still experimental - See below</a>)</td>
+    <td>Stable<br>Experimental Features:
```
Comment (bpaduch) [inline extensions/amp-form/amp-form.md]: Rewrite as follows:
Stable with the following <a href="https://www.ampproject.org/docs/reference/experimental.html">Experimental</a> features:

File: extensions/amp-form/amp-form.md [hunk 3]
```diff
@@ -291,6 +296,38 @@ This shows all validation errors on all invalid inputs when the form is submitte
...
+```html
+<!-- Initial Load -->
+<form ...>
+  <input name="canonicalUrl" type="hidden" default-value="The canonical URL is: CANONICAL_URL">
+  <input name="clientId" type="hidden" default-value="CLIENT_ID(myid)">
+  ...
+</form>
+```
+
+Once the user tries to submit the form, AMP will try to resolve the variables and update the fields values. For XHR submissions, all variables are likely to be substituted and resolved. However, in non-XHR GET submissions, values that requires async-resolution might not be available and might not be resolved if it has not previously been resolved. `CLIENT_ID` for example would not resolve if it wasn't resolved and cached previously.
```
Comment (bpaduch) [inline extensions/amp-form/amp-form.md]: change "fields" to "fields'"

File: extensions/amp-form/amp-form.md (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: I am happy to change this, `default-value` was a quick suggestion from @cramforce and I took it and ran with it 😄. 

> (a) Is using data- attributes a better approach here. (I'm not sure where to use them and when not.)

Me and you both. I am never sure whether to choose to go with `data-` or not. I'd love for us to decide on a better way to choose and set of rules to follow when evaluating attributes naming.

> (b) "default" suggests to me something that's defined (in fact, it's the value used if another one doesn't come along to replace it). Here, however, what's really being constructed is a tokenized string. I wonder if we can make the attribute name clearer.

We could rename `default-value` to `resolve-value` or something clearer, value-for-substitution also sounds weird but might be clearer.

File: extensions/amp-form/amp-form.md [hunk 4]
```diff
@@ -291,6 +296,38 @@ This shows all validation errors on all invalid inputs when the form is submitte
 #### As You Go
 This allows your user to see validation messages as they're interacting with the input, if the email they typed is invalid they'll see the error right away and once fixed the error goes away.
 
+## Variable Substitutions
+__(<a href="https://www.ampproject.org/docs/reference/experimental.html">experimental</a>)__
+
+`amp-form` allows [variable substitutions](../../spec/amp-var-substitutions.md) for inside hidden inputs with `default-value` attribute. On each submission, `amp-form` finds all `input[type=hidden]` inside the form and apply variable substitutions to `default-value` attribute and fill the result in the input `value`.
+
+Here's an example of how inputs are before and after substitutions:
```
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: This only exposes platform variables, I believe analytics variables are only exposed through analytics. Will update to reflect.

File: extensions/amp-form/amp-form.md (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Yeah sure, would this look like the following:
```html
<input value="can: CANONICAL_URL, rand: RANDOM" data-amp-replace="CANONICAL_URL RANDOM">
```

`amp-form` would then call a modified `maybeExpandLink` with the expected attribute to replace (`value` in this case and `href` in links case).

@rudygalfi @cramforce does this sound better?

Comment (cramforce) [inline extensions/amp-form/amp-form.md]: SG

Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: LGTM. Should `data-amp-replace` be a comma separated list?

Comment (cramforce) [inline extensions/amp-form/amp-form.md]: I think we should change it to be space separated because that is more idiomatic on the web.

Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Sent out #6621 to update the format we're doing here. PTAL there and I'll update docs PR after we get it in.

Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

**Comment by mkhatib** [comment]
PTAL 👀

File: extensions/amp-form/amp-form.md (hunk 2 above)
Comment (bpaduch) [inline extensions/amp-form/amp-form.md]: Can you add the link to https://www.ampproject.org/docs/reference/experimental for the word "Experimental".

File: extensions/amp-form/amp-form.md (hunk 1 above)
Comment (bpaduch) [inline extensions/amp-form/amp-form.md]: It doesn't look like the wording changed.  Can you change to what I wrote above.

File: extensions/amp-form/amp-form.md [hunk 5]
```diff
@@ -23,7 +23,12 @@ limitations under the License.
   </tr>
   <tr>
     <td width="40%"><strong>Availability</strong></td>
-    <td>Stable<br>(<a href="#custom-validations">Custom Validation still experimental - See below</a>)</td>
+    <td>Stable with the following Experimental features:
+       <ul>
+          <li><a href="#custom-validations">Custom Validation</a></li>
+          <li><a href="#variable-substitution">Variable Substitutions</a></li>
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: missing s: variable-substitutions

File: extensions/amp-form/amp-form.md [hunk 6]
```diff
@@ -291,6 +296,46 @@ This shows all validation errors on all invalid inputs when the form is submitte
 #### As You Go
 This allows your user to see validation messages as they're interacting with the input, if the email they typed is invalid they'll see the error right away and once fixed the error goes away.
 
+## Variable Substitutions
+__(<a href="https://www.ampproject.org/docs/reference/experimental.html">experimental</a>)__
+
+`amp-form` allows [platform variable substitutions](../../spec/amp-var-substitutions.md) for inputs that are hidden and that have the `data-amp-replace` attribute. On each form submission, `amp-form` finds all `input[type=hidden][data-amp-replace]` inside the form and applies variable substitutions to its `value` attribute and replaces it with the result of the substitution.
+
+It's important to provide the variables you are using for each substitution on each input through a space-separated strings in `data-amp-replace`, see examples below. AMP will not replace variables that are not explicitly provided.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: Can this be cleaned up to: You must provide the variables you are using for each substitution on each input by specifying a space-separated string of the variables used in `data-amp-replace` (see example below). AMP will not replace variables that are not explicitly specified.

File: extensions/amp-form/amp-form.md (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

**Comment by rudygalfi** [comment]
LGTM, thanks!

**Pull request merged**